### DISCUSSION TITLE
Making semicolons compulsory in Mindcode

### DISCUSSION THREAD

**Discussion by cardillan** [cardillan/mindcode#120] [Polls]
I'm planning some enhancement to Mindcode syntax in, hopefully, not so distant future.  A rough scope of planned enhancements can be found in the [roadmap](https://github.com/cardillan/mindcode/blob/main/ROADMAP.markdown#updating-antlr-grammar). All the planned keywords will probably become reserved, although the planned functionality can take some time to materialize. In any case, it can be expected that the changes will break some existing Mindcode.

This would provide an opportunity to make one far-reaching change to Mindcode: to make semicolons compulsory as separators between expressions. Most of existing Mindcode doesn't utilize semicolons, so consequences of this change would be far reaching and would require a lot of updates.

Optional semicolons were inspired by Ruby, but unfortunately they allow a lot of ambiguities in source code. Some of the problematic constructs include:

- `set a = 10`: compiles as two expressions, `set` and `a = 10`. `set` is an expression that evaluates the variable called `set`, but because it isn't used for anything, it is just silently ignored. It might lead users to believe that `set variable = expression` is a regular Mindcode syntax.
- `printflush message1`: compiles as two expressions, `printflush` and `message1`. Similarly to the previous case, both of them are just silently ignored. Users might be baffled why it doesn't appear in the compiled code, reducing the confidence in the compiler.
- `Lorem ipsum dolor sit amet consectetur adipiscing elit`: again, no error reported on this "source code", but nothing gets generated in the compiled code. Gives Mindcode a somewhat Dadaist touch.
- `print(a)`: amazingly, the current syntax allows to interpret this construct as two expressions, `print` and `(a)`. ANTLR fortunately always prefers the correct interpretation, that is a call to a `print` function with `a` as an argument. Compulsory semicolons would make the former interpretation impossible, which would be very desired.

So, I'm addressing you, dear Mindcode users, with the following question:

**Comment by limonovthesecond2** [comment]
Another example (if I understood everything correctly):
```
if x == 1
    print("1");
elif x = 2
    print("2");
end
```

- Semicolons are optional: the compiler, after successful compilation, will at most warn about unused variable `elif`. `x = 2` won't bother him at all.
- Semicolons are required: the compiler simply won't compile it and report that the keyword `elif` was not found as the operation `x = 2 print("2");`.

**Reply by limonovthesecond2** [reply]
Just a question, since this is a discussion: can a semicolon be replaced with a newline?

**Reply by cardillan** [reply]
That's a possibility, but has a significant drawback: it becomes necessary to escape line endings if you have expressions spanning several lines. Such expressions aren't very common in Mindcode in my experience - but they can occur.

If you're familiar with Visual Basic for Applications - the macro language used in Microsoft Office - you might remember that this language uses underscore to escape line endings. I don't like it very much.

**Reply by cardillan** [reply]
Regarding the example above: yes, this is about right. 

The error message in the second case wouldn't be about unknown `elif` keyword - the parser doesn't know you wanted to use a keyword at that place (it disregards any indents). I'm using the ANTLR4 parser, which tries to match the source code to any possible syntax construct, and when no such a match is possible, displays a list of symbols that would allow the matching to continue (it is a state machine, so it just displays a list of allowable states at that point). So, in this case, it would probably say it expects one out of a huge list of symbols after the `elif`, e.g. `;`, all the operators (`+`, `-`, `*` and so on), even the property access operator `.`, maybe an opening bracket and so on. These syntax error messages aren't very intuitive, but I have no idea how to improve them.

**Comment by cardillan** [comment]
For some time, I was considering an idea of defining additional grammar which could then be even less compatible with the current one (modeled probably after Java instead of Ruby). That way users could rewrite their code for the new syntax if/when they would feel like it. The two grammars would support the same constructs, only would express them in different ways.

There's a lot of problems with it. I would have to support the two grammars, meaning either twice as much work for very little gain, or letting the old grammar to become obsolete. And since I'd switch to the new grammar immediately, I wouldn't be able to spot problems with the older one unless someone reported it. Finally, neither the Mindcode project, nor the community around it seems large enough to afford such an extravaganza.

**Comment by cardillan** [comment]
Thanks to all that have participated in the poll. Given the majority opinion, I've decided to make semicolons compulsory in the new syntax, which is being discussed in #142. Since the issue is resolved, I'm closing this thread - if you have anything to add, please use the new thread describing all the planned changes.
